lib/vnet/daemon/client_darwin.go

```diff
@@ -0,0 +1,74 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//go:build darwin
+// +build darwin
+
+package daemon
+
+// #cgo CFLAGS: -Wall -xobjective-c -fblocks -fobjc-arc -mmacosx-version-min=10.15
+// #cgo LDFLAGS: -framework Foundation -framework ServiceManagement
+// #include "client_darwin.h"
+import "C"
+
+import (
+	"context"
+	"errors"
+	"os/exec"
+	"unsafe"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+var log = logutils.NewPackageLogger(teleport.ComponentKey, "vnet:daemon")
+
+// Taken from manual for codesign.
+const codesignExitCodeVerificationFailed = 1
+
+func IsSigned(ctx context.Context) (bool, error) {
+	var result C.BundlePathResult
+	defer func() {
+		C.free(unsafe.Pointer(result.bundlePath))
+	}()
+	C.BundlePath(&result)
+
+	bundlePath := C.GoString(result.bundlePath)
+	if bundlePath == "" {
+		return false, nil
+	}
+
+	if err := exec.CommandContext(ctx, "codesign", "--verify", bundlePath).Run(); err != nil {
```

Comment:
`os.Executable` wouldn't work here – it has to be the path to the bundle itself, otherwise `codesign --verify` returns what I'd consider false positives. For example, if you run it on `build/tsh`, it tells you that everything is A-okay. And that's true – `build/tsh` is not in any app bundle, which means there's no signing requirements placed upon it, so technically it fulfills all requirements.

Reply:
`codesign --verify` doesn't work when tsh is executed from a symlink in `/usr/local/bin`. This can happen when tsh is installed through tsh.app.

In that scenario, `C.BundlePath` returns `/usr/local/bin`, which has a similar effect as calling `codesign --verify` on `build` in the teleport repo.

I'll need to switch to `spctl --assess`, which AFAIR works with individual items. So instead of finding the bundle path, we'll be able to call it directly on the tsh executable from `Contents/MacOS`.

Reply:
`spctl --assess` fails on unnotarized bundles, which means running it on `tshdev/tsh.app/Contents/MacOS/tsh` that's signed locally won't work.

Reply:
I think I'll keep using `codesign --verify`, but instead of using Obj-C to get the bundle path, I'll resolve any symlinks to `os.Executable`, then check if the dir path ends with `Contents/MacOS` and then just remove that suffix.

This check isn't done for security, it's merely to know whether we're running from inside a signed package or if we should run the osascript prompt.